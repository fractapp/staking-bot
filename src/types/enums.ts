export enum Network {
    Polkadot = 0,
    Kusama,
}

export enum Action {
    Open,
    AddAmount,
    CreateWithdrawRq,
    ConfirmWithdraw,
    UpdateNominations
}


export enum Currency {
    DOT = 0,
    KSM,
}


export function getNativeCurrency(network: Network): Currency {
    switch (network) {
        case Network.Polkadot:
            return Currency.DOT
        case Network.Kusama:
            return Currency.KSM
    }
}

export function getNetwork(currency: Currency): Network {
    switch (currency) {
        case Currency.DOT:
            return Network.Polkadot
        case Currency.KSM:
            return Network.Kusama
    }
}

export function getSymbol(currency: Currency) {
    let symbol = '';
    switch (currency) {
        case Currency.DOT:
            symbol = 'DOT';
            break;
        case Currency.KSM:
            symbol = 'KSM';
            break;
        default:
            throw new Error('invalid currency');
    }
    return symbol;
}

export function toCurrency(currency: string): Currency {
    switch (currency) {
        case 'DOT':
            return Currency.DOT;
        case 'KSM':
            return Currency.KSM;
        default:
            throw new Error('invalid currency');
    }
}

export function fromCurrency(currency: Currency): string {
    switch (currency) {
        case Currency.DOT:
            return 'DOT';
        case Currency.KSM:
            return 'KSM';
        default:
            throw new Error('invalid currency');
    }
}
