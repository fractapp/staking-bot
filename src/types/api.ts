import {Currency} from "./enums";

export type UndeliveredMessagesInfo = {
    messages: Array<Message>,
    users: {
        [id in string]: Profile
    }
};
export type Profile = {
    id: string;
    name: string;
    username: string;
    avatarExt: string;
    lastUpdate: number;
    addresses: {
        [id in Currency]: string
    };
    isChatBot: boolean;
};

export type MessageRq = {
    version: number,
    value: string,
    action: string,
    receiver: string,
    args: {
        [key in string]: string
    },
    rows: Array<Row>
};

export type Message = {
    id: string;
    value: string;
    action: string | null;
    args: {
        [key in string]: string
    },
    rows: Array<Row>;
    timestamp: number;
    sender: string;
    receiver: string;
    hideBtn: boolean;
};

export type Row = {
    buttons: Array<Button>
};

export type Button = {
    value: string
    action: string
    arguments: {
        [key in string]: string
    }
    imageUrl: string | null,
};

export enum DefaultMsgAction {
    Init = "/init",
    OpenUrl = "/openUrl",
    EnterAmount = "/enterAmount",
    Broadcast = "/broadcast",
    WalletButtonOut = '/walletOut',
    WalletButtonIn = '/walletIn'
}
