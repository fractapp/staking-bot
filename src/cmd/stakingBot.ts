import {DefaultMsgAction, Message, Profile} from "../types/api";
import {MsgAction, Network, toCurrency} from "../types/enums";
import {ApiPromise, WsProvider} from "@polkadot/api";
import dotenv from "dotenv";
import {ActionHandler} from "../stakingBot/actions";
import {FractappClient} from "../fractapp/client";
import {DB} from "../db/db";
import {Scheduler} from "../stakingBot/scheduler";
import {CacheClient} from "../utils/cacheClient";
import {Const} from "../utils/const";

dotenv.config()

const clientUrl = process.env["FRACTAPP_CLIENT_URL"] as string
const client = new FractappClient(clientUrl)
let actionHandler: ActionHandler

let cache: Map<Network, CacheClient> = new Map<Network, CacheClient>()

async function start() {
    const connectionString = process.env["MONGODB_CONNECTION"] as string
    const cacheServerPort = process.env["CACHE_PORT"] as string

    const seed = process.env["SEED"] as string
    const apiPolkadot = await getApiInstance(process.env["POLKADOT_RPC_URL"] as string)
    const apiKusama = await getApiInstance(process.env["KUSAMA_RPC_URL"] as string)

    cache.set(Network.Polkadot, new CacheClient(cacheServerPort, Network.Polkadot))
    cache.set(Network.Kusama, new CacheClient(cacheServerPort, Network.Kusama))

    const db = new DB(connectionString)

    console.log("connect to db...")
    await db.connect()
    console.log("end of connect to db")

    actionHandler = new ActionHandler(client, apiPolkadot, apiKusama, cache.get(Network.Polkadot)!, cache.get(Network.Kusama)!, db)
    const scheduler = new Scheduler(cache.get(Network.Polkadot)!, cache.get(Network.Kusama)!, db, actionHandler)

    console.log("init blockchain info...")

    let isInit = false
    while (!isInit) {
        const polkadotCache = cache.get(Network.Polkadot)!
        const kusamaCache = cache.get(Network.Kusama)!

        console.log("init cache")

        try {
            await polkadotCache.updateCache()
            await kusamaCache.updateCache()

            await actionHandler.init()

            isInit = true
        } catch (e) {
            console.log("wait init cache")
        }

        await new Promise(resolve => setTimeout(resolve, 5 * Const.Sec));
    }

    console.log("end of init blockchain info")

    console.log("fractapp authorization...")
    await client.auth(seed)
    console.log("fractapp authorization: Success")

    setInterval(async () => {
        console.log("start scheduler")
        await scheduler.call()
        console.log("end of scheduler")
    }, Const.SchedulerTimeout)

    console.log("start app...")
    const usersPromise: Map<string, Promise<void>> = new Map<string, Promise<void>>()

    while (true) {
        try {
            const newMsgs = await client.getUnreadMessages()

            for (const msg of newMsgs.messages) {
                console.log("msg id: " + msg.id)
                console.log("msg: " + JSON.stringify(msg))

                if (!usersPromise.has(msg.sender)) {
                    try {
                        const p = action(msg, newMsgs.users[msg.sender]).finally(() => usersPromise.delete(msg.sender))
                        usersPromise.set(msg.sender, p)
                    } catch (e) {
                        usersPromise.delete(msg.sender)
                        console.log("Error: " + e)
                    }
                }
            }
        } catch (e) {
            console.log("Error: " + e)
        }
        await new Promise(resolve => setTimeout(resolve, 1000));
    }
}

async function getApiInstance(wssUrl: string): Promise<ApiPromise> {
    const wsProvider = new WsProvider(wssUrl);
    return await ApiPromise.create({provider: wsProvider})
}

async function action(msg: Message, user: Profile) {
    //TODO: error count for msg
    try {
        switch (msg.action) {
            case DefaultMsgAction.Init:
                await actionHandler.initMsg(msg, user)
                break
            case MsgAction.ChooseNewDeposit:
                await actionHandler.chooseNewDeposit(msg, user)
                break
            case MsgAction.Enter:
                await actionHandler.enter(toCurrency(msg.args.currency), Number(msg.args.action), msg.sender, user)
                break
            case MsgAction.ChooseWithdraw:
                await actionHandler.chooseWithdraw(msg)
                break
            case MsgAction.Confirm:
                await actionHandler.confirm(msg, user)
                break
            case MsgAction.SuccessTx:
                await actionHandler.successTx(msg)
                break
            case MsgAction.ErrorTx:
                await actionHandler.errorTx(msg)
                break
            case DefaultMsgAction.WalletButtonIn:
            case MsgAction.MyDeposits:
                await actionHandler.myDeposits(msg, user)
                break
            case DefaultMsgAction.WalletButtonOut:
            case MsgAction.WithdrawRequests:
                await actionHandler.withdrawRequests(msg, user)
                break
            default:
                await actionHandler.initMsg(msg, user)
        }
    } catch (e) {
        await actionHandler.initMsg(msg, user)
    }
    await client.setRead([msg.id])
    console.log("Success: (msgId)" + msg.id)
}

start()
