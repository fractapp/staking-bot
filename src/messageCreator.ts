import {Button, Message, Profile} from "./types/api";
import {Action, Currency, fromCurrency, Network} from "./types/enums";
import {MathUtil} from "./math";
import {ApiPromise} from "@polkadot/api";
import {Cache} from "./cache";
import {Const} from "./const";
import {MsgAction} from "./actions";
import BN from "bn.js";

export type DepositInfo = {
    value: string | null,
    hasWithdrawRequest: boolean,
    hasActiveAmount: boolean,
    isWarnExist: boolean
}

export type NewDepositInfo = {
    value: string,
    hasOpenedDeposit: boolean
}

export type WithdrawRequestsInfo = {
    value: string | null,
    totalWithdrawPlanks: BN,
    buttons: Array<Button>
}

export class MessageCreator {
    private readonly api: ApiPromise
    private readonly cache: Cache
    private readonly network: Network
    private readonly currency: Currency

    constructor(api: ApiPromise, cache: Cache, network: Network, currency: Currency) {
        this.api = api
        this.cache = cache
        this.network = network
        this.currency = currency
    }

    public warning(activeAmount: BN): string | null {
        const cacheInfo = this.cache.getStakingInfo()
        const amount = MathUtil.convertFromPlanckToString(cacheInfo.minStakingAmountPlanks, cacheInfo.decimalsCount)

        if (activeAmount.cmp(cacheInfo.minStakingAmountPlanks) < 0) {
            return `\n\nThe $${fromCurrency(this.currency)} deposit is not active. The deposit must have more than the minimum amount of ${amount} ${fromCurrency(this.currency)}!`
        }

        return null
    }
    public async getNewDeposit(user: Profile): Promise<NewDepositInfo> {
        const info = this.cache.getStakingInfo()
        const min = info.minStakingAmountPlanks.mul(Const.Accuracy).div(info.decimals).toNumber() / Const.Accuracy.toNumber()
        const staking = await this.api.query.staking.ledger(user.addresses[this.currency])
        const stakingInfo = staking.isNone ? null : staking.unwrap()

        const value =
            (this.currency == Currency.KSM ? 'ðŸ”µKSM DepositðŸ”µ\n' : 'ðŸ”´DOT DepositðŸ”´\n') +
            `Annual yield: ${info.averageAPY}%\n` +
            `Withdrawal duration: ${info.unbondingPeriod} days \n` +
            `Minimum amount: ${min} ${fromCurrency(this.currency)}`

        return {
            value: value,
            hasOpenedDeposit: stakingInfo != null && stakingInfo.total.toBn().cmp(new BN(0)) != 0
        }
    }

    public async getDeposit(user: Profile): Promise<DepositInfo> {
        const staking = await this.api.query.staking.ledger(user.addresses[this.currency])
        const cacheInfo = this.cache.getStakingInfo()

        const stakingInfo = staking.isNone ? null : staking.unwrap()

        let value = null
        let isWarnExist = false
        if (stakingInfo != null && stakingInfo.total.toBn().cmp(new BN(0)) != 0) {
            const warn = this.warning(stakingInfo.active.toBn())
            if (warn != null) {
                isWarnExist = true
            }

            value =   (this.currency == Currency.KSM ? 'ðŸ”µKSM DepositðŸ”µ\n' : 'ðŸ”´DOT DepositðŸ”´\n') +
                `Withdrawal requests: ${stakingInfo!.unlocking.length}\n` +
                `Active amount: ${MathUtil.convertFromPlanckToString(stakingInfo!.active.toBn(), cacheInfo.decimalsCount)} ${fromCurrency(this.currency)}\n` +
                `Withdrawal amount: ${MathUtil.convertFromPlanckToString(stakingInfo!.total.toBn().sub(stakingInfo!.active.toBn()), cacheInfo.decimalsCount)} ${fromCurrency(this.currency)}` +
                (warn ?? "")
        }

        return {
            value: value,
            hasWithdrawRequest: stakingInfo != null && stakingInfo!.unlocking.length > 0,
            hasActiveAmount: stakingInfo != null && stakingInfo!.active.toBn().cmp(new BN(0)) > 0,
            isWarnExist: isWarnExist
        }
    }

    public async getWithdrawRequests(user: Profile): Promise<WithdrawRequestsInfo> {
        const staking = await this.api.query.staking.ledger(user.addresses[this.currency])

        const info = this.cache.getStakingInfo()
        const stakingInfo = staking.isNone ? null : staking.unwrap()

        const buttons = []
        let totalWithdrawPlanks = new BN(0)
        let withdrawRequestText = ""
        if (stakingInfo != null && stakingInfo.unlocking.length > 0) {
            const activeEra = await this.api.query.staking.activeEra()

            withdrawRequestText = this.currency == Currency.DOT ? `ðŸ”´DOT requestsðŸ”´\n\n` : `ðŸ”µKSM requestsðŸ”µ\n\n`

            let isExistNotActiveWR = false
            for (let unlocking of stakingInfo.unlocking) {
                const planksValue = unlocking.value.unwrap().toBn()
                const amount = MathUtil.convertFromPlanckToString(planksValue, info.decimalsCount)
                const era = unlocking.era.unwrap().toNumber()
                const duration = era - activeEra.unwrap().index.toBn().toNumber()
                const time = Math.floor(duration * info.eraDuration) / 24

                if (time < 0) {
                    totalWithdrawPlanks = totalWithdrawPlanks.add(planksValue)
                } else {
                    if (!isExistNotActiveWR) {
                        isExistNotActiveWR = true
                    }
                    withdrawRequestText += `${amount} ${fromCurrency(this.currency)} unlocked after ${time} days\n`
                }
            }

            if (!isExistNotActiveWR) {
                withdrawRequestText = this.warning(stakingInfo.active.toBn()) ?? ""
            } else {
                withdrawRequestText += this.warning(stakingInfo.active.toBn()) ?? ""
            }
        }

        let value: string | null = null
        if (withdrawRequestText) {
            value = withdrawRequestText
        }

        if (totalWithdrawPlanks.cmp(new BN(0)) > 0) {
            const withdrawText = `${MathUtil.convertFromPlanckToString(totalWithdrawPlanks, info.decimalsCount)} ${fromCurrency(this.currency)}`
            value += `\nYou can withdraw ${withdrawText}`
            buttons.push({
                value: `Withdraw ${withdrawText}`,
                action: MsgAction.Confirm,
                arguments: {
                    arguments: JSON.stringify({
                        currency: fromCurrency(this.currency),
                        action: String(Action.ConfirmWithdraw)
                    })
                },
                imageUrl: null
            })
        }

        if (stakingInfo != null && stakingInfo.active.toBn().cmp(new BN(0)) > 0) {
            buttons.push({
                value: `Create withdraw request for ${fromCurrency(this.currency)}`,
                action: MsgAction.Enter,
                arguments: {
                    currency: fromCurrency(this.currency),
                    action: String(Action.CreateWithdrawRq)
                },
                imageUrl: null
            })
        }

        return {
            value: value,
            totalWithdrawPlanks: totalWithdrawPlanks,
            buttons: buttons,
        }
    }
}
