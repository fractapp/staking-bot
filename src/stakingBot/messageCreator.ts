import {Button, Profile} from "../types/api";
import {Action, Currency, fromCurrency, MsgAction, Network} from "../types/enums";
import {MathUtil} from "../utils/math";
import {Const} from "../utils/const";
import {CacheClient} from "../utils/cacheClient";

export type DepositInfo = {
    value: string | null,
    hasWithdrawRequest: boolean,
    hasActiveAmount: boolean,
    isWarnExist: boolean
}

export type NewDepositInfo = {
    value: string,
    hasOpenedDeposit: boolean
}

export type WithdrawRequestsInfo = {
    value: string | null,
    totalWithdrawPlanks: bigint,
    buttons: Array<Button>
}

export class MessageCreator {
    private readonly cache: CacheClient
    private readonly network: Network
    private readonly currency: Currency

    constructor(cache: CacheClient, network: Network, currency: Currency) {
        this.cache = cache
        this.network = network
        this.currency = currency
    }

    public async warning(activeAmount: bigint): Promise<string | null> {
        const cacheInfo = await this.cache.getStakingInfo()

        if (activeAmount < cacheInfo.minStakingAmountPlanks) {
            const minAmount = MathUtil.convertFromPlanckToString(cacheInfo.minStakingAmountPlanks, cacheInfo.decimalsCount)
            return `\n\nThe ${fromCurrency(this.currency)} deposit is not active. The deposit must have more than the minimum amount of ${minAmount} ${fromCurrency(this.currency)}!`
        }

        return null
    }

    public async getNewDeposit(user: Profile): Promise<NewDepositInfo> {
        const info = await this.cache.getStakingInfo()
        const usersStaking = await this.cache.getUsersStaking()
        const min = Number(info.minStakingAmountPlanks * Const.Accuracy / info.decimals) / Number(Const.Accuracy)
        const staking = usersStaking[user.addresses[this.currency]]
        const stakingInfo = staking == undefined ? null : staking

        const value =
            (this.currency == Currency.KSM ? 'ðŸ”µKSM DepositðŸ”µ\n' : 'ðŸ”´DOT DepositðŸ”´\n') +
            `Annual yield: ${info.averageAPY}%\n` +
            `Withdrawal duration: ${info.unbondingPeriod} days \n` +
            `Minimum amount: ${min} ${fromCurrency(this.currency)}`

        return {
            value: value,
            hasOpenedDeposit: stakingInfo != null && stakingInfo.total != 0n
        }
    }

    public async getDeposit(user: Profile): Promise<DepositInfo> {
        const cacheInfo = await this.cache.getStakingInfo()
        const usersStaking = await this.cache.getUsersStaking()

        const staking = usersStaking[user.addresses[this.currency]]
        const stakingInfo = staking == undefined ? null : staking

        let value = null
        let isWarnExist = false
        if (stakingInfo != null && stakingInfo.total != 0n) {
            const warn = await this.warning(stakingInfo.active)
            if (warn != null) {
                isWarnExist = true
            }

            value = (this.currency == Currency.KSM ? 'ðŸ”µKSM DepositðŸ”µ\n' : 'ðŸ”´DOT DepositðŸ”´\n') +
                `Total withdrawal request: ${stakingInfo!.unlocking.length}\n` +
                `Active deposit amount: ${MathUtil.convertFromPlanckToString(stakingInfo!.active, cacheInfo.decimalsCount)} ${fromCurrency(this.currency)}\n` +
                `Withdrawal amount: ${MathUtil.convertFromPlanckToString(stakingInfo!.total - stakingInfo!.active, cacheInfo.decimalsCount)} ${fromCurrency(this.currency)}` +
                (warn ?? "")
        }

        return {
            value: value,
            hasWithdrawRequest: stakingInfo != null && stakingInfo!.unlocking.length > 0,
            hasActiveAmount: stakingInfo != null && stakingInfo!.active > 0,
            isWarnExist: isWarnExist
        }
    }

    public async getWithdrawRequests(user: Profile): Promise<WithdrawRequestsInfo> {
        const info = await this.cache.getStakingInfo()
        const usersStaking = await this.cache.getUsersStaking()

        const staking = usersStaking[user.addresses[this.currency]]
        const stakingInfo = staking == undefined ? null : staking

        const buttons = []
        let totalWithdrawPlanks = 0n
        let withdrawRequestText = ""
        if (stakingInfo != null && stakingInfo.unlocking.length > 0) {
            const activeEra = info.activeEra

            withdrawRequestText = this.currency == Currency.DOT ? `ðŸ”´DOT requestsðŸ”´\n\n` : `ðŸ”µKSM requestsðŸ”µ\n\n`

            let isExistNotActiveWR = false
            for (let unlocking of stakingInfo.unlocking) {
                const planksValue = unlocking.value
                const amount = MathUtil.convertFromPlanckToString(planksValue, info.decimalsCount)
                const duration = unlocking.era - activeEra
                const time = Math.floor(duration * info.eraDuration) / 24

                if (time < 0) {
                    totalWithdrawPlanks += planksValue
                } else {
                    if (!isExistNotActiveWR) {
                        isExistNotActiveWR = true
                    }
                    withdrawRequestText += `${amount} ${fromCurrency(this.currency)} unlocked after ${time} days\n`
                }
            }

            if (!isExistNotActiveWR) {
                withdrawRequestText = await this.warning(stakingInfo.active) ?? ""
            } else {
                withdrawRequestText += await this.warning(stakingInfo.active) ?? ""
            }
        }

        let value: string | null = null
        if (withdrawRequestText) {
            value = withdrawRequestText
        }

        if (totalWithdrawPlanks > 0n) {
            const withdrawText = `${MathUtil.convertFromPlanckToString(totalWithdrawPlanks, info.decimalsCount)} ${fromCurrency(this.currency)}`
            value += `\n\nYou can withdraw ${withdrawText}`
            buttons.push({
                value: `Withdraw ${withdrawText}`,
                action: MsgAction.Confirm,
                arguments: {
                    arguments: JSON.stringify({
                        currency: fromCurrency(this.currency),
                        action: String(Action.ConfirmWithdraw)
                    })
                },
                imageUrl: null
            })
        }

        if (stakingInfo != null && stakingInfo.active > 0n) {
            buttons.push({
                value: `Create withdraw request for ${fromCurrency(this.currency)}`,
                action: MsgAction.Enter,
                arguments: {
                    currency: fromCurrency(this.currency),
                    action: String(Action.CreateWithdrawRq)
                },
                imageUrl: null
            })
        }

        return {
            value: value,
            totalWithdrawPlanks: totalWithdrawPlanks,
            buttons: buttons,
        }
    }
}
