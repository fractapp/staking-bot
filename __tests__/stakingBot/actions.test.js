import {ActionHandler} from "../../src/stakingBot/actions";
import {TxBuilder} from "../../src/utils/txBuilder";
import {MessageCreator} from "../../src/stakingBot/messageCreator";
import {Action, Currency, fromCurrency} from "../../src/types/enums";
import {Profile} from "../../src/types/api";

jest.mock('@polkadot/api', () => ({}))
jest.mock('../../src/fractapp/client', () => ({}))
jest.mock('../../src/db/db', () => ({}))
jest.mock('../../src/utils/txBuilder', () => ({
    TxBuilder: jest.fn()
}))
jest.mock('../../src/stakingBot/messageCreator', () => ({
    MessageCreator: jest.fn()
}))

const api = {
    rpc: {
        payment: {
            queryInfo: jest.fn()
        }
    }
};
const cache = {
    getStakingInfo: jest.fn(),
    getUsersStaking: jest.fn()
}
const txBuilder = {
    bound: jest.fn(),
    fakeSign: jest.fn(),
    updateNominations: jest.fn()
}
const messageCreator = {
    getDeposit: jest.fn(),
    getNewDeposit: jest.fn(),
    getWithdrawRequests: jest.fn()
}
const client = {
    sendMsg: jest.fn()
}
const db = {
    hasUser: jest.fn(),
    setUser: jest.fn()
}
const sender = "sender"
const user = {
    id: "id",
    addresses: ["1", "2"]
}

MessageCreator.mockReturnValue(messageCreator)

it('test init', async () => {
    const initMock = jest.fn()
    TxBuilder.mockReturnValue({
        init: initMock
    })

    const action = new ActionHandler(client, api, api, cache, cache, db)
    await action.init()

    expect(initMock).toBeCalledTimes(2)
});

it('test initMsg', async () => {
    cache.getStakingInfo.mockReturnValueOnce({
        averageAPY: 13
    })
    cache.getStakingInfo.mockReturnValueOnce({
        averageAPY: 6
    })

    messageCreator.getDeposit.mockReturnValueOnce({
        value: "Polkadot deposit",
        hasWithdrawRequest: true,
        hasActiveAmount: false,
        isWarnExist: true
    })
    messageCreator.getDeposit.mockReturnValueOnce({
        value: "Kusama deposit",
        hasWithdrawRequest: false,
        hasActiveAmount: true,
        isWarnExist: true
    })

    const action = new ActionHandler(client, api, api, cache, cache, db)
    await action.initMsg({
        sender: sender
    },  user)

    expect(client.sendMsg.mock.calls[0]).toMatchSnapshot()
});

it('test chooseNewDeposit', async () => {
    messageCreator.getNewDeposit.mockReturnValueOnce({
        value: "Polkadot deposit",
        hasOpenedDeposit: true
    })
    messageCreator.getNewDeposit.mockReturnValueOnce({
        value: "Kusama deposit",
        hasOpenedDeposit: false
    })

    const action = new ActionHandler(client, api, api, cache, cache, db)
    await action.chooseNewDeposit({
        sender: sender
    },  user)

    expect(client.sendMsg.mock.calls[0]).toMatchSnapshot()
})

it('test myDeposits', async () => {
    messageCreator.getDeposit.mockReturnValueOnce({
        value: "Polkadot deposit",
        hasWithdrawRequest: true,
        hasActiveAmount: false,
        isWarnExist: true
    })
    messageCreator.getDeposit.mockReturnValueOnce({
        value: "Kusama deposit",
        hasWithdrawRequest: false,
        hasActiveAmount: true,
        isWarnExist: true
    })

    const action = new ActionHandler(client, api, api, cache, cache, db)
    await action.myDeposits({
        sender: sender
    },  user)

    expect(client.sendMsg.mock.calls[0]).toMatchSnapshot()
})

it('test withdrawRequests', async () => {
    messageCreator.getWithdrawRequests.mockReturnValueOnce({
        value: "Polkadot",
        totalWithdrawPlanks: 1000n,
        buttons: [
            {
                value: "v1",
                action: "action",
                arguments: {},
                imageUrl: null
            }
        ]
    })
    messageCreator.getWithdrawRequests.mockReturnValueOnce({
        value: "Kusama",
        totalWithdrawPlanks: 3000n,
        buttons: [
            {
                value: "v2",
                action: "action",
                arguments: {},
                imageUrl: null
            }
        ]
    })

    const action = new ActionHandler(client, api, api, cache, cache, db)
    await action.withdrawRequests({
        sender: sender
    },  user)

    expect(client.sendMsg.mock.calls[0]).toMatchSnapshot()
})

it('test chooseWithdraw', async () => {
    const action = new ActionHandler(client, api, api, cache, cache, db)
    await action.chooseWithdraw({
        sender: sender
    })

    expect(client.sendMsg.mock.calls[0]).toMatchSnapshot()
})

it('test successTx', async () => {
    const action = new ActionHandler(client, api, api, cache, cache, db)
    await action.successTx({
        sender: sender
    })

    expect(client.sendMsg.mock.calls[0]).toMatchSnapshot()
})

it('test errorTx', async () => {
    const action = new ActionHandler(client, api, api, cache, cache, db)
    await action.errorTx({
        sender: sender
    })

    expect(client.sendMsg.mock.calls[0]).toMatchSnapshot()
})

it('test enter', async () => {
    TxBuilder.mockReturnValue(txBuilder)

    cache.getUsersStaking.mockReturnValueOnce({})

    cache.getStakingInfo.mockReturnValueOnce({
        minStakingAmountPlanks: BigInt(150)
    })

    const unsignedTx = "unsignedTx"
    txBuilder.bound.mockReturnValueOnce(unsignedTx)
    api.rpc.payment.queryInfo.mockReturnValueOnce({
        partialFee: {
            toBn: () => 10n
        }
    })
    const action = new ActionHandler(client, api, api, cache, cache, db)
    await action.enter(Currency.DOT, Action.Open, sender, user, "error")

    expect(client.sendMsg.mock.calls[0]).toMatchSnapshot()
})
