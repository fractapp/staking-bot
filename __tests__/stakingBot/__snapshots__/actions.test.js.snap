// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test chooseNewDeposit 1`] = `
Array [
  Object {
    "action": "",
    "args": Object {},
    "receiver": "sender",
    "rows": Array [
      Object {
        "buttons": Array [
          Object {
            "action": "enter",
            "arguments": Object {
              "action": "1",
              "currency": "DOT",
            },
            "imageUrl": null,
            "value": "üî¥ Add amount to DOT deposit",
          },
        ],
      },
      Object {
        "buttons": Array [
          Object {
            "action": "enter",
            "arguments": Object {
              "action": "0",
              "currency": "KSM",
            },
            "imageUrl": null,
            "value": "üîµ Open KSM deposit",
          },
        ],
      },
    ],
    "value": "Deposits information:

Polkadot deposit

Kusama deposit",
    "version": 1,
  },
]
`;

exports[`test chooseWithdraw 1`] = `
Array [
  Object {
    "action": "",
    "args": Object {},
    "receiver": "sender",
    "rows": Array [
      Object {
        "buttons": Array [
          Object {
            "action": "enter",
            "arguments": Object {
              "action": "2",
              "currency": "DOT",
            },
            "imageUrl": null,
            "value": "üî¥ DOT",
          },
          Object {
            "action": "enter",
            "arguments": Object {
              "action": "2",
              "currency": "KSM",
            },
            "imageUrl": null,
            "value": "üîµ KSM",
          },
        ],
      },
    ],
    "value": "Please, choose deposit for withdraw: ",
    "version": 1,
  },
]
`;

exports[`test enter 1`] = `
Array [
  Object {
    "action": "",
    "args": Object {},
    "receiver": "sender",
    "rows": Array [
      Object {
        "buttons": Array [
          Object {
            "action": "/enterAmount",
            "arguments": Object {
              "arguments": "{\\"currency\\":\\"DOT\\",\\"action\\":\\"0\\"}",
              "currency": "DOT",
              "fee": "10",
              "limit": "",
              "next": "confirm",
            },
            "imageUrl": null,
            "value": "üíµ Enter amount",
          },
          Object {
            "action": "/init",
            "arguments": Object {},
            "imageUrl": null,
            "value": "‚úñÔ∏è Cancel",
          },
        ],
      },
    ],
    "value": "error",
    "version": 1,
  },
]
`;

exports[`test errorTx 1`] = `
Array [
  Object {
    "action": "",
    "args": Object {},
    "receiver": "sender",
    "rows": Array [
      Object {
        "buttons": Array [
          Object {
            "action": "/init",
            "arguments": Object {},
            "imageUrl": null,
            "value": "‚Ü©Ô∏è Try again",
          },
        ],
      },
    ],
    "value": "An error has occurred. Please, try again or contact with support: support@fractapp.com.",
    "version": 1,
  },
]
`;

exports[`test initMsg 1`] = `
Array [
  Object {
    "action": "",
    "args": Object {},
    "receiver": "sender",
    "rows": Array [
      Object {
        "buttons": Array [
          Object {
            "action": "chooseNewDeposit",
            "arguments": Object {},
            "imageUrl": null,
            "value": "Open deposit",
          },
          Object {
            "action": "/openUrl",
            "arguments": Object {
              "link": "https://wiki.polkadot.network/ru-RU/docs/learn-staking",
            },
            "imageUrl": null,
            "value": "Info",
          },
        ],
      },
      Object {
        "buttons": Array [
          Object {
            "action": "myDeposits",
            "arguments": Object {},
            "imageUrl": null,
            "value": "My deposits",
          },
        ],
      },
    ],
    "value": "Hi!
You can open a staking deposit for DOT or KSM and receive from 6% to 13% per annum!

Also, you have already opened deposits.

Your deposits are inactive because it is below the minimum amount.",
    "version": 1,
  },
]
`;

exports[`test myDeposits 1`] = `
Array [
  Object {
    "action": "",
    "args": Object {},
    "receiver": "sender",
    "rows": Array [
      Object {
        "buttons": Array [
          Object {
            "action": "enter",
            "arguments": Object {
              "action": "2",
              "currency": "KSM",
            },
            "imageUrl": null,
            "value": "Create a withdrawal request",
          },
        ],
      },
      Object {
        "buttons": Array [
          Object {
            "action": "chooseNewDeposit",
            "arguments": Object {},
            "imageUrl": null,
            "value": "Add amount",
          },
        ],
      },
      Object {
        "buttons": Array [
          Object {
            "action": "withdrawRequests",
            "arguments": Object {},
            "imageUrl": null,
            "value": "Withdrawal requests",
          },
        ],
      },
      Object {
        "buttons": Array [
          Object {
            "action": "/init",
            "arguments": Object {},
            "imageUrl": null,
            "value": "Back",
          },
        ],
      },
    ],
    "value": "Your deposits: 

Polkadot deposit

Kusama deposit

",
    "version": 1,
  },
]
`;

exports[`test successTx 1`] = `
Array [
  Object {
    "action": "",
    "args": Object {},
    "receiver": "sender",
    "rows": Array [
      Object {
        "buttons": Array [
          Object {
            "action": "/init",
            "arguments": Object {},
            "imageUrl": null,
            "value": "‚Ü©Ô∏è Return to menu",
          },
        ],
      },
    ],
    "value": "Wait for the confirmation of the transaction.",
    "version": 1,
  },
]
`;

exports[`test withdrawRequests 1`] = `
Array [
  Object {
    "action": "",
    "args": Object {},
    "receiver": "sender",
    "rows": Array [
      Object {
        "buttons": Array [
          Object {
            "action": "action",
            "arguments": Object {},
            "imageUrl": null,
            "value": "v1",
          },
        ],
      },
      Object {
        "buttons": Array [
          Object {
            "action": "action",
            "arguments": Object {},
            "imageUrl": null,
            "value": "v2",
          },
        ],
      },
      Object {
        "buttons": Array [
          Object {
            "action": "myDeposits",
            "arguments": Object {},
            "imageUrl": null,
            "value": "Back",
          },
        ],
      },
    ],
    "value": "Your withdrawal requests: 

Polkadot
Kusama",
    "version": 1,
  },
]
`;
